src/main/org/deidentifier/arx/ShadowModelMembershipRisk.java

```diff
@@ -336,30 +376,627 @@ private Data getCopy(DataHandle handle, int[] sample) {
+                    IntIntOpenHashMap map = new IntIntOpenHashMap();
+                    for (int row = 0; row < handle.getNumRows(); row++) {
+                        int code = handle.internalGetEncodedValue(row, column, false); // Beware that code can be -1
+                        map.putOrAdd(code, 1, 1);
+                    }
+                    
+                    int[] availableValues = metaData.getAvialableValues(i);
+                    double[] freqs = new double[availableValues.length];
+                    
+                    // Count freqs of all available values.
+                    for(int v = 0; v < availableValues.length; v++) {
+                        int value = availableValues[v];
+                        if(map.containsKey(value)) {
+                            freqs[v] = map.get(availableValues[v]);
+                        } else {
+                            freqs[v] = 0d;
+                        }
+                    }
+                    
+                    result[i] = freqs;
+                    //System.out.println(Arrays.toString(freqs));
+                }
+                System.out.println(attributeName + " --> " + Arrays.toString(result[i]));
+            }
+            // flatten array
+            double[] flatResult = flattenArray(result);
+            //System.out.println(Arrays.toString(flatResult));
+            return flatResult;
+        }
+        /**
+         * Calculates correlation features using Pearson's product-moment correlation.
+         * All columns of continuous attributes are directly used for correlation calculation.
+         * Categorical and ordinal attributes are transfered to a sparse representation
+         * were each value becomes an own column and whether (or not) the value applies to a row
+         * is indicated by the value 1d (or 0d).
+         * 
+         * @return
+         */
+        //TODO: Ass support for ALL attribute Types
+        private double[] calculateCorrelationFeatures() {
+            
+            // Initialize dataframe used to store the parts of the input matrix to assemble
+            List<List<Double>> frame = new ArrayList<List<Double>>();
+            
+            for (int i = 0; i < columns.length; i++) {
+
+                // Obtain attribute details
+                int c = columns[i];
+                String attributeName = handle.getAttributeName(c);
+                DataType<?> _type = handle.getDefinition().getDataType(attributeName);
+                Class<?> _clazz = _type.getDescription().getWrappedClass();
+                
+                if (_clazz.equals(Double.class)) {
+                    
+                    // Directly add column to frame
+                    frame.add(Arrays.asList(getColumnAsDouble(c)));
+                    
+                } else if (_clazz.equals(String.class)) {
+                    
+                    //TODO replace with more efficient Code !!!11
```

Comment:
Short answer: It’s how they do it in the Python code.
Long(er) answer: I assume that a Pearson correlation between variables only makes sense when there is a linear relationship between the variables. As an example, one can consider age as continuous attribute and several diseases as categorical attribute encoded as follows: {0 – Chickenpox; 1 – Alzheimer’s disease; 2 – Cancer; 3 - Measles}. Although there is a relationship between age and the diseases mentioned, it could not be captured due to the random assignment of labels. This can be avoided by splitting the presence of each diseases into an own attribute. This would be my “non-specialist” explanation of the issue.

Reply:
Makes sense.